contracts/interchain-token-service/src/contract/execute.rs

```diff
@@ -217,169 +255,158 @@ pub fn set_chain_config(
 
-        // Token chain info must already exist from previous deployment
-        ensure!(
-            token_chain_info.is_some(),
-            Error::TokenNotDeployed {
-                token_id: message.token_id(),
-                chain: message_direction.clone().into(),
-            }
-        );
+    if let Some(GlobalTokenConfig { origin_chain, .. }) = token_config {
+        ensure_origin_matches_source_chain(source_chain, origin_chain, token_id.clone())?;
     } else {
-        // Token is being deployed for the first time
-        let token_config = TokenConfig {
-            origin_chain: ChainNameRaw::from(message_direction.clone()),
-        };
+        initialize_token_on_origin(storage, source_chain, token_id.clone())?;
+    }
 
-        state::save_token_config(storage, &message.token_id(), &token_config)
-            .change_context(Error::State)?;
+    let instance = TokenInstance::new(token_id.clone(), deployment_type);
 
-        ensure!(
-            token_chain_info.is_none(),
-            Error::TokenAlreadyDeployed {
-                token_id: message.token_id(),
-                chain: message_direction.clone().into(),
-            }
-        );
-    }
+    state::save_token_instantiation(storage, destination_chain, token_id, &instance)
+        .change_context(Error::State)
+}
+
+fn ensure_origin_matches_source_chain(
+    source_chain: ChainNameRaw,
+    origin_chain: ChainNameRaw,
+    token_id: TokenId,
+) -> Result<(), Error> {
+    ensure!(
+        origin_chain == source_chain,
+        Error::TokenDeployedFromNonOriginChain {
+            token_id,
+            origin_chain,
+            chain: source_chain.clone(),
+        }
+    );
 
     Ok(())
 }
 
-/// Ensures that the token is not being redeployed to the same destination chain.
-fn token_redeployment_check(
-    message: &Message,
-    message_direction: &MessageDirection,
-    token_chain_info: &Option<TokenChainInfo>,
+fn initialize_token_on_origin(
+    storage: &mut dyn Storage,
+    source_chain: ChainNameRaw,
+    token_id: TokenId,
 ) -> Result<(), Error> {
-    if matches!(
-        message,
-        Message::DeployInterchainToken { .. } | Message::DeployTokenManager { .. }
-    ) && matches!(message_direction, MessageDirection::To(_))
-        && token_chain_info.is_some()
-    {
-        bail!(Error::TokenAlreadyDeployed {
-            token_id: message.token_id(),
-            chain: message_direction.clone().into(),
-        });
-    }
+    // Token is being deployed for the first time
+    let token_config = GlobalTokenConfig {
+        token_id: token_id.clone(),
+        origin_chain: source_chain.clone(),
+    };
+    let instance = TokenInstance::new_on_origin(token_id.clone());
 
+    state::save_global_token_config(storage, &token_id, &token_config)
+        .and_then(|_| state::save_token_instantiation(storage, source_chain, token_id, &instance))
```

Comment:
I'd rather see these as separate statements for readability and since the ordering is independent

Reply:
because I need to change context on both statement otherwise I didn't find it more readable the other way around, but I don't feel strongly about it